nca/CoreDS/ConnectivityProperties.py

```diff
@@ -0,0 +1,587 @@
+            assert peers_for_named_ports
+            assert not port_set_with_named_ports.port_set
+            assert len(port_set_with_named_ports.named_ports) == 1
+            port = list(port_set_with_named_ports.named_ports)[0]
+            for peer in peers_for_named_ports:
+                real_ports = ConnectivityProperties.resolve_named_port(port, peer, protocols)
+                if not real_ports:
+                    continue
+                if src_ports.named_ports:
+                    props = ConnectivityProperties(source_ports=real_ports, dest_ports=dst_ports,
+                                                   protocols=protocols, methods=methods,
+                                                   paths=paths_dfa, hosts=hosts_dfa,
+                                                   icmp_type=icmp_type, icmp_code=icmp_code,
+                                                   src_peers=base_peer_set.get_peer_interval_of(PeerSet({peer})),
+                                                   dst_peers=dst_peers_interval, base_peer_set=base_peer_set)
+                else:
+                    props = ConnectivityProperties(source_ports=src_ports, dest_ports=real_ports,
+                                                   protocols=protocols, methods=methods,
+                                                   paths=paths_dfa, hosts=hosts_dfa,
+                                                   icmp_type=icmp_type, icmp_code=icmp_code,
+                                                   src_peers=src_peers_interval,
+                                                   dst_peers=base_peer_set.get_peer_interval_of(PeerSet({peer})),
+                                                   base_peer_set=base_peer_set)
+                conn_properties |= props
+        return conn_properties
+
+    @staticmethod
+    def make_conn_props_from_dict(peer_container, cube_dict):
+        """
+        Create ConnectivityProperties from the given cube
+        :param PeerContainer peer_container: the set of all peers
+        :param dict cube_dict: the given cube represented as a dictionary
+        :return: ConnectivityProperties
+        """
+        cube_dict_copy = cube_dict.copy()
+        src_ports = cube_dict_copy.pop("src_ports", PortSet(True))
+        dst_ports = cube_dict_copy.pop("dst_ports", PortSet(True))
+        protocols = cube_dict_copy.pop("protocols", ProtocolSet(True))
+        src_peers = cube_dict_copy.pop("src_peers", None)
+        dst_peers = cube_dict_copy.pop("dst_peers", None)
+        paths_dfa = cube_dict_copy.pop("paths", None)
+        hosts_dfa = cube_dict_copy.pop("hosts", None)
+        methods = cube_dict_copy.pop("methods", MethodSet(True))
+        icmp_type = cube_dict_copy.pop("icmp_type", DimensionsManager().get_dimension_domain_by_name('icmp_type'))
+        icmp_code = cube_dict_copy.pop("icmp_code", DimensionsManager().get_dimension_domain_by_name('icmp_code'))
```

Comment:
we have many places (at least 3) where we duplicate the hard-coded semantics of defaults for a missing dimension in a cube input. can we avoid this duplication?

Reply:
Generalized.

Reply:
It is still duplicated in the `__init__` and `make_conn_props` default values for each argument, as well as in the `get_full_dim_having_orig_type` implementation.

Reply:
How about storing a constant dict with all defaults and use dict.update() to override non-default values?